lib/notifications/display.dart

```diff
@@ -92,7 +93,35 @@ class NotificationDisplayManager {
   static Future<void> _onMessageFcmMessage(MessageFcmMessage data, Map<String, dynamic> dataJson) async {
     assert(debugLog('notif message content: ${data.content}'));
     final zulipLocalizations = GlobalLocalizations.zulipLocalizations;
-    final title = switch (data.recipient) {
+    final groupKey = _groupKey(data);
+    final conversationKey = _conversationKey(data, groupKey);
+
+    final oldMessagingStyle = await ZulipBinding.instance.androidNotificationHost
+      .getActiveNotificationMessagingStyleByTag(conversationKey);
+
+    final MessagingStyle messagingStyle;
+    if (oldMessagingStyle != null) {
+      messagingStyle = oldMessagingStyle;
+      messagingStyle.messages =
+        oldMessagingStyle.messages.toList(); // Clone fixed-length list to growable.
+    } else {
+      messagingStyle = MessagingStyle(
+        user: Person(
+          key: _personKey(data.realmUri, data.userId),
+          name: zulipLocalizations.notifSelfUser),
+        messages: [],
+        isGroupConversation: switch (data.recipient) {
+          FcmMessageStreamRecipient() => true,
+          FcmMessageDmRecipient(:var allRecipientIds) when allRecipientIds.length > 2 => true,
+          FcmMessageDmRecipient() => false,
+        });
+    }
+
+    // For stream messages, the title remains constant. For DMs, it may
```

Comment:
Hmm, why do you say it remains constant? It can change if the channel/stream gets renamed, right?

I think the comment in zulip-mobile is accurate here too. It just needs "PM" modernized to read "DM", and the reference to https://github.com/zulip/zulip-mobile/issues/5116 can refer to the existing TODO below… or better yet, file a zulip-flutter issue for better titling group DM threads in notifications, referring back to https://github.com/zulip/zulip-mobile/issues/5116 .